📖 Add MachinePool controller architecture docs (#4003)
kubernetes-sigs/cluster-api: docs/book/src/developer/architecture/controllers/machine-pool.md

> # Machine  Controller

Author: whoops, good catch

![](../../../images/cluster-admission-machinepool-controller.png)

The Machine controller's main responsibilities are:

> * Setting an OwnerReference on:
>     * Each MachinePool object to the Cluster object.
>     * The associated BootstrapConfig object.
>     * The associated InfrastructureMachinePool object.

Reviewer: ```suggestion
* Setting an OwnerReference on each MachinePool object to:
    * The associated Cluster object.
    * The associated BootstrapConfig object.
    * The associated InfrastructureMachinePool object.
```

* Copy data from `BootstrapConfig.Status.DataSecretName` to `MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` if
`MachinePool.Spec.Template.Spec.Bootstrap.DataSecretName` is empty.

> * Setting NodeRefs to be able to associate instances and kubernetes nodes.

Reviewer: Are these two the same? Just wanted to clarify where the NodeRefs are being set and what instances were we referring to.
```suggestion
* Setting NodeRefs on MachinePool instances to be able to associate them with kubernetes nodes.
```

> * Deleting Nodes in the target cluster when the associated instance is deleted.

Reviewer: nit: Again just clarifying.
```suggestion
* Deleting Nodes in the target cluster when the associated MachinePool instance is deleted.
```

> * Cleanup of related objects.

Reviewer: What are the related objects? Is it referring to things like BootstrapConfig?

Author: removed, this doesn't apply here. There is one bootstrap config per machine pool (not one per instance)

### Bootstrap provider

The BootstrapConfig object **must** have a `status` object.

To override the bootstrap provider, a user (or external system) can directly set the `MachinePool.Spec.Bootstrap.DataSecretName`
field. This will mark the machine as ready for bootstrapping and no bootstrap data secret name will be copied from the
BootstrapConfig object.

#### Required `status` fields

The `status` object **must** have several fields defined:

* `ready` - a boolean field indicating the bootstrap config data is generated and ready for use.
* `dataSecretName` - a string field referencing the name of the secret that stores the generated bootstrap data.

#### Optional `status` fields

The `status` object **may** define several fields that do not affect functionality if missing:

* `failureReason` - a string field explaining why a fatal error has occurred, if possible.
* `failureMessage` - a string field that holds the message contained by the error.

Example:

```yaml
kind: MyBootstrapProviderConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
status:
    ready: true
    dataSecretName: "MyBootstrapSecret"
```

### Infrastructure provider

> The InfrastructureMachine object **must** have both `spec` and `status` objects.

Reviewer: Should this be InfrastructureMachinePool? 
For example, I noticed that the `AWSMachinePoolSpec` has `ProviderIDList` whereas the `AWSMachineSpec` only has a `ProviderID` field.
```suggestion
The InfrastructureMachinePool object **must** have both `spec` and `status` objects.
```

#### Required `spec` fields

The `spec` object **must** have at least one field defined:

> * `providerIDList` - the list of cloud provider IDs identifying the instances.

Reviewer: > identifying the instances.

Are these machine instances or node instances? I'm guessing they're the same.

Author: they're provider instances, for example ASG instances or VMSS instances. 1 instance = 1 node

#### Required `status` fields

The `status` object **must** have at least one field defined:

* `ready` - a boolean field indicating if the infrastructure is ready to be used or not.

#### Optional `status` fields

The `status` object **may** define several fields that do not affect functionality if missing:

* `failureReason` - is a string that explains why a fatal error has occurred, if possible.
* `failureMessage` - is a string that holds the message contained by the error.

Example:
```yaml
kind: MyMachinePool
apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
spec:
    providerIDList: 
      - cloud:////my-cloud-provider-id-0
      - cloud:////my-cloud-provider-id-1
status:
    ready: true
```

### Secrets

The machine pool controller will use a secret in the following format:

| secret name | field name | content |
|:---:|:---:|---|

> |`<cluster-name>-kubeconfig`|`value`|base64 encoded kubeconfig that is authenticated with the child cluster|

Reviewer: On line 14 we have _target cluster_,
on line 17,24 we have _workload cluster_.

Suggesting we use _workload cluster_ everywhere since that's what we have in the glossary.

```suggestion
|`<cluster-name>-kubeconfig`|`value`|base64 encoded kubeconfig that is authenticated with the workload cluster|
```

To override the bootstrap provider, a user (or external system) can directly set the `MachinePool.Spec.Bootstrap.DataSecretName`
field. This will mark the machine as ready for bootstrapping and no bootstrap data secret name will be copied from the
BootstrapConfig object.

#### Required `status` fields

The `status` object **must** have several fields defined:

* `ready` - a boolean field indicating the bootstrap config data is generated and ready for use.
* `dataSecretName` - a string field referencing the name of the secret that stores the generated bootstrap data.

#### Optional `status` fields

The `status` object **may** define several fields that do not affect functionality if missing:

* `failureReason` - a string field explaining why a fatal error has occurred, if possible.
* `failureMessage` - a string field that holds the message contained by the error.

Example:

```yaml
kind: MyBootstrapProviderConfig
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
status:
    ready: true
    dataSecretName: "MyBootstrapSecret"
```

### Infrastructure provider

The InfrastructureMachine object **must** have both `spec` and `status` objects.

#### Required `spec` fields

> The `spec` object **must** at least one field defined:

Reviewer 2: ```suggestion
The `spec` object **must** have at least one field defined:
```

* `providerIDList` - the list of cloud provider IDs identifying the instances.

#### Required `status` fields

> The `status` object **must** at least one field defined:

Reviewer 2: ```suggestion
The `status` object **must** have at least one field defined:
```
